### DISCUSSION TITLE
Using array params in dynamic tasks

### DISCUSSION THREAD

**Discussion by pilord12** [apache/airflow#33707] [Q&A] [answered]
I am trying to create a DAG which can take lists of inputs and map each element to an Operator execution.  As an example, one such input list would be some S3 buckets to be created:

```
with DAG(
        dag_id="provision-fileset",
        schedule=None,
        start_date=pendulum.datetime(2023, 8, 1),
        params={
            "create_bucket_names": Param(
                type="array",
                description="Name of all buckets to be created",
                default=[],
            )
        },
        render_template_as_native_obj=True
) as my_dag:
    b = S3CreateBucketOperator.partial(task_id="create_bucket", aws_conn_id="localstack_dev",
                                       dag=my_dag).expand(bucket_name="{{ params.create_bucket_names }}")
```

However, the above code gives an exception `S3CreateBucketOperator.expand() got an unexpected type 'str' for keyword argument bucket_name`.  What is the appropriate way to reference an array parameter in `.expand`?  Using Airflow 2.6.3 if that makes a difference.

**Comment by potiuk** [comment]
I think you need to set https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/operators.html#rendering-fields-as-native-python-objects to True. Otherwise string processed by JINJA will remain string.

**Reply by pilord12** [reply]
That is already set to true--see the above code snippet in the DAG definition :)

**Reply by potiuk** [reply]
Ah. You can't. You need to set the parameter as array not string. I have not noticed you are just passing string to expand. You need to convert the string to list. in python. There is no jinja template processing haapening on regular strings being set as parameter - this is only happening on  attributes of tasks.

**Reply by potiuk** [reply]
And you can't do it this way either. It makes no sens to run expand  this way - you can pass it an array, or output of another task (then the output will be expanded to array when the previous task executes). Right now you are just passing it "{{ xxx}}" string that makes no sense

**Reply by pilord12** [reply]
> You need to convert the string to list. in python. 

Right.  I also tried `bucket_name="{{ params.create_bucket_names }}".split("\n")` but got `TypeError: expected string or bytes-like object` coming from the S3 operator.  Is there a more idiomatic way to be doing this?

> output of another task

What is the best way to do this?  I tried using a Python task to grab the list parameter, and attempted a couple things which didn't work:

* Setting a global variable in Python - only passed the initial empty array value to the `.expand`, presumably because the `.expand` is evaluated eagerly even if its resulting tasks depend on the Python task.
* Setting an Airflow variable - failed for similar reasons

Would you recommend XComs?  I am a bit confused based on [this documentation](https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/xcoms.html), I'm not sure I understand where the `task_instance` that is referenced comes from.

**Reply by potiuk** [reply, marked as answer]
You should have a task that takes the parameter you pass, convert it to array and produces the array as output, and then use that ouput as input to your task and  that will be expanded.

**Reply by pilord12** [reply]
Ah, was totally unaware that you could directly use the return value of Python tasks as input to other tasks--I thought you had to pass from one task to another using XComs or something.  The below code worked for me.  Thanks for the guidance!

```
with DAG(
        dag_id="provision-fileset",
        schedule=None,
        start_date=pendulum.datetime(2023, 8, 1),
        params={
            "create_bucket_names": Param(
                type="array",
                description="Name of all buckets to be created",
                default=[],
            )
        },
        render_template_as_native_obj=True
) as my_dag:
    @task
    def extract_params():
        context = get_current_context()
        return context["params"]["create_bucket_names"]

    a = extract_params()

    b = S3CreateBucketOperator.partial(task_id="create_bucket", aws_conn_id="localstack_dev",
                                       dag=my_dag).expand(bucket_name=a)
```

**Reply by potiuk** [reply]
Glad I could help . Yes. That's the whole point with inputs/outputs of task flow - it almost feels like you are running regular Python methods and take their return values as parameters. It's not exactly this and has a few caveats but it feels mostly like this

**Reply by pilord12** [reply]
I guess this makes me a little confused on the use case for XComs.  I'm sure I don't fully understand them, but it seems like you could just store values in Task results and reference them whenever necessary in other Tasks instead of needing to push and pull from XCom storage.

**Reply by potiuk** [reply]
>  but it seems like you could just store values in Task results and reference them whenever necessary in other Tasks instead of needing to push and pull from XCom storage.

No. I think you understand it well, but you have not realized that the `@task` method results .... use XComs in fact. Task flow decorators are really syntactic sugar to do whatever you are doing with regular operators with Xcom. 

In short when you return result from `@task` decorated method, it is .... pushed to XCom (by the `@task` decorator). And when you use result of such `@task` decorated method you actually get not the actual result - but `XComArg` representation of the XCom saved result (also`@task` decorator will take care about that). The XComArg gets lazily evaluated to `pull` the result from XCom when their value is used inside the function that use it.

So by using `@task` decorators you are using XCom, but it's done transparently for you without manually doing all the push/pull dance.

You can read more about it here https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/taskflow.html#taskflow

**Reply by pilord12** [reply]
Ahh, that makes perfect sense.  Much more convenient than XCom push/pull 😄 I wasn't aware of the history of the TaskFlow API--seems like it's been added comparatively recently and isn't so much providing specific functionality but more making it easier to use existing functionality.

**Reply by potiuk** [reply]
It is pretty old one in fact - like 3 years :)

**Reply by pilord12** [reply]
Just goes to show how new Airflow is to me 😅

**Reply by potiuk** [reply]
:D

**Reply by misterprakhar** [reply]
> Ah. You can't. You need to set the parameter as array not string. I have not noticed you are just passing string to expand. You need to convert the string to list. in python. There is no jinja template processing haapening on regular strings being set as parameter - this is only happening on attributes of tasks.

If we need to convert it to list ourselves, then what is the use of `render_template_as_native_obj=True`? Shouldn't it render it as a list if the type is set as array when defining the params??
